### PR TITLE
Added `OrderedDictionary<TKey, TValue>` with nullable key support, #11

### LINKED ISSUE #11: Create lower-level implementation of LinkedDictionary<TKey, TValue>
The `LinkedDictionary<TKey, TValue>` is a composite type that internally is using a `LinkedList<T>` and `Dictionary<TKey, TValue>`. This is a stopgap to get the behavior we need, but we could achieve better performance by using a lower-level hashtable structure to manage the a single collection of items rather than having 2 collections.

The main differences between `Dictionary<TKey, TValue>` and `LinkedDictionary<TKey, TValue>` are:

- `LinkedDictionary<TKey, TValue>` maintains insertion order across edits
- `LinkedDictionary<TKey, TValue>` allows a `null` key

### PR DESCRIPTION
**Pull request by paulirwin**
This PR adds the OrderedDictionary type with nullable key support, as well as unit tests.

Fixes #11

### CODE REVIEW

**Review by NightOwl888 — CHANGES_REQUESTED** [review]
Thanks for the PR.

This isn't a full review, but I thought I would leave some early feedback. I skimmed the tests, but took a good look at the `OrderedDictionary` implementation and it looks pretty solid. There are a few ideas we can borrow from the other collections, though. See the inline comments.

The doc comments look light, but it appears that this is one of those new features that Microsoft hasn't fully documented yet. Generally, we grab everything except for the code examples to put into the docs and of course point out how our implementation differs from Microsoft's in the header of the class. We also outline all of the features that are similar to Java. In this case, we would want to point out that this is similar to a `LinkedHashMap<K, V>` in the JDK except for it is missing the constructor overload to change it into an LRU cache. `LurchTable<TKey, TValue> is what we use for that currently, although it is a pretty heavy implementation. Have a look at the `LinkedDictionary<TKey, TValue>` header, since this class should have the same behavior.

Also, I noticed a lot of places where we are using `throw` inline (and calling properties of the `SR` class, which allocate strings on the stack). Ideally, we would move to `ThrowHelper`, but it is less important for those non-public methods, especially in the non-generic interfaces. Lucene.NET uses the generic interfaces a lot, though, so it would be good to make those overloads as optimal as we can.

File: src/J2N/ThrowHelper.cs [hunk 1]
```diff
@@ -941,6 +943,30 @@ internal static void IfNullAndNullsAreIllegalThenThrow<T>(object? value, Excepti
...
 
+        internal static void ThrowIfNegative(int value, ExceptionArgument argument)
+        {
+            if (value < 0)
+            {
+                ThrowArgumentOutOfRange_MustBeNonNegative(argument);
+            }
+        }
+
+        internal static void ThrowIfNull(object? value, ExceptionArgument argument)
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/J2N/ThrowHelper.cs]: The ThrowIfNull removal left warnings behind that I'm not sure how to fix. We could just remove those checks and let null fail elsewhere, but let me know what you think.

File: src/J2N/Collections/Generic/EqualityComparer.cs [hunk 2]
```diff
@@ -28,6 +28,59 @@ namespace J2N.Collections.Generic
...
+        /// </summary>
+        /// <param name="equals">The delegate to use to implement the <see cref="IEqualityComparer{T}.Equals(T,T)"/> method.</param>
+        /// <param name="getHashCode">
+        /// The delegate to use to implement the <see cref="IEqualityComparer{T}.GetHashCode(T)"/> method.
+        /// If no delegate is supplied, calls to the resulting comparer's <see cref="IEqualityComparer{T}.GetHashCode(T)"/>
+        /// will throw <see cref="NotSupportedException"/>.
+        /// </param>
+        /// <returns>The new comparer.</returns>
+        /// <exception cref="ArgumentNullException">The <paramref name="equals"/> delegate was null.</exception>
+        public static IEqualityComparer<T> Create(Func<T?, T?, bool> equals, Func<T, int>? getHashCode = null)
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/J2N/Collections/Generic/EqualityComparer.cs]: Moved to TestCommon as it was only used in the tests.

File: src/J2N/Collections/Generic/OrderedDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,1902 @@
...
+        /// <param name="key">The key to insert.</param>
+        /// <param name="value">The value to insert.</param>
+        /// <exception cref="ArgumentNullException">key is null.</exception>
+        /// <exception cref="ArgumentException">An element with the same key already exists in the <see cref="OrderedDictionary{TKey, TValue}"/>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="index"/> is less than 0 or greater than <see cref="Count"/>.</exception>
+        public void Insert(int index, [AllowNull] TKey key, TValue value)
+        {
+            if ((uint)index > (uint)_count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRange_IndexMustBeLessOrEqualException();
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/J2N/Collections/Generic/OrderedDictionary.cs:737]: There wasn't a ThrowIndexArgumentOutOfRange method, so I found the closest one that let the tests pass. But I'll add that method and change this so it matches better.

File: tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    public abstract class OrderedDictionary_Generic_Tests<TKey, TValue> : IDictionary_Generic_Tests<TKey, TValue>
+    {
+        #region IDictionary<TKey, TValue> Helper Methods
+
+        protected override bool DefaultValueAllowed => true;
+        protected override bool DefaultValueWhenNotAllowed_Throws => false;
+        protected override bool Enumerator_Empty_UsesSingletonInstance => true;
+        protected override bool Enumerator_Empty_Current_UndefinedOperation_Throws => true;
+        protected override bool Enumerator_Empty_ModifiedDuringEnumeration_ThrowsInvalidOperationException => false;
+        protected override ModifyOperation ModifyEnumeratorThrows => ModifyOperation.Add | ModifyOperation.Insert | ModifyOperation.Remove | ModifyOperation.Clear;
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:26]: Are you sure that is needed? LinkedDictionary does not support this. If needed, I'll probably let you take that change.

[the remark this answers is not in the document]
Comment (NightOwl888) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:26]: The change is rather trivial. The only thing required is to remove the `_version++;` line from `RemoveAt()` and `Clear()`. In both cases, the `MoveNext()` method will not throw the exception and will skip over any removed values.

On the other hand, working around the problem from outside of the class is very costly - we would have to store the deleted keys in another collection and then remove them after the enumeration is complete or fall back on using LINQ for enumeration.

In Java all iterators have a `remove()` method, so if this is used anywhere in our codebase now or in the future, it is ideal to support it.

File: src/J2N/ThrowHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NightOwl888) [reply src/J2N/ThrowHelper.cs]: I think I zigged when I should have zagged here. I didn't realize Microsoft was conditionally compiling.

https://github.com/dotnet/runtime/blob/v9.0.4/src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs#L9-L14

Since that is the case, then adding backward support in `ThrowHelper` and calling the new static methods on the exceptions is the right call. At least for this method. The other two I could go either way.

Adding conditional compilation will engage the intrinsics that may be built in (depending on hardware) in .NET 8+.

File: tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulirwin) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:26]: I looked into your suggestion and I don't believe that approach (removing the version check for Remove and Clear) would be correct in all cases. If the enumerator's `_index` is equal to `dictionary._count - 1` before the call to MoveNext, for example, then if an item earlier in the collection's entry array is removed, then it will get to the `if (_index < dictionary._count)` condition, which is no longer true (because `_count` was decremented by one, so `_index` is now `== dictionary._count`), and so it will fail to move next to the last item, and not return it from the enumerator. Of course this is exacerbated by Clear or multiple Remove calls during enumeration. As noted in the remarks on `_entries`, this version does not leave holes that can just be skipped over when removed.

So I do not think we can safely just ignore the version check for Remove or Clear to solve this. Likewise, since this array is accessed by index, Add or Insert during enumeration could cause it to return wrong/duplicated values, although you only mentioned Remove, Clear, and modifying. Modifying the value should be fine so I can make that change to the test assertions easily enough.

[the remark this answers is not in the document]
Comment (NightOwl888) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:26]: Good call. It is more involved than I thought. But there isn't much extra state. There are just 2 extra variables to track, `_freeCount` and `_freeList`, which are both `int`. The most complicated bit is the fact that `Entry.Next` has 2 purposes in `Dictionary<TKey, TValue>` instead of a single purpose. If it is less than -1, then it represents a free entry in a specific order, which is then reused in that order.

https://github.com/dotnet/runtime/blob/v8.0.15/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1355-L1359

Maintaining these three values requires some rework of the `TryInsert`, `RemoveAt`, `Clear`, and `MoveNext`, but it looks like Microsoft stayed closely aligned enough to the `Dictionary<TKey, TValue>` design so that it won't be too painful to achieve.

Note that in `Dictionary<TKey, TValue>`, the code to find a `null` key was duplicated to eliminate the `comparer` in the loop for performance reasons. Ideally, the remove logic would be in place before we do that optimization.

https://github.com/NightOwl888/J2N/blob/v2.1.0/src/J2N/Collections/Generic/Dictionary.cs#L1051

File: src/J2N/Collections/Generic/OrderedDictionary.cs [hunk 5]
```diff
@@ -0,0 +1,2125 @@
...
+        public void RemoveAt(int index)
+        {
+            int count = _count;
+            if ((uint)index >= (uint)count)
+            {
+                ThrowIndexArgumentOutOfRange();
+            }
+
+            // Remove from the associated bucket chain the entry that lives at the specified index.
+            RemoveEntryFromBucket(index);
```
Comment (NightOwl888) [inline src/J2N/Collections/Generic/OrderedDictionary.cs:814]: I noticed a bug that we might want to report upstream. In `Dictionary<TKey, TValue>`, when a bucket is cleared it also clears the values if there are any reference types involved so the GC can collect them. 

https://github.com/dotnet/runtime/blob/v8.0.15/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1010-L1018

But it appears that `OrderedDictionary<TKey, TValue>` will hold onto these references until the bucket is reused or the collection is GC'd. Since Java only uses reference types, this will be a problem we will run into frequently.

File: src/J2N/ThrowHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/J2N/ThrowHelper.cs]: Updated; let me know what you think.

File: tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulirwin) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:26]: It sounds like you have a better understanding of this than I do; I've reviewed that code and don't completely follow it. I'd appreciate it if you could take this PR from here and finish it up if there are any other things you'd like done. It is set to allow edits from maintainers.

[the remark this answers is not in the document]
Comment (NightOwl888) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:26]: Alright. Go ahead and finish up the PR without the enumerator patch and I'll make that change in another PR. It would be good to have a stable place to revert to if it ends up going sideways. I suspect the ordering is always filling up the beginning of the indexing array (after all, the loop would fail if there were a break) and it would be possible to continue ordering the free space, but it may not end up working out that way.

File: src/J2N/Collections/Generic/OrderedDictionary.cs (hunk 5 above)
Comment (paulirwin) [reply src/J2N/Collections/Generic/OrderedDictionary.cs:810]: I initially agreed with your assessment that that's how the code seems, however, when I wrote an app to try to reproduce this with large arrays as the values before filing the bug upstream, I found it was actually properly GC'ing it in both the J2N and .NET 9 versions. After reading the code again, I figured out where that was happening: at the `entries![--_count] = default;` line below this one after it shifts the entries. So I'll remove my "fix" and leave it be. It is possible when this is converted to use "holes," however, that you will need to take a different approach.

**Review by NightOwl888 — CHANGES_REQUESTED** [review]
This is looking good.

I found some issues to address to fit it more in line with the JDK-like expectations, particularly some missing members and tests for `null` keys. Please see the inline comments.

### Links to Upstream Sources

Microsoft's code is frequently being updated and it would not be practical for us to do a complete comparison of all components during each release. So, recently I have been adding a direct link to the upstream source file including the tag for the release at the top of each file so it is easy to determine how far out of date the file is. As mentioned, normally this is done with the release tag. I never link to a branch because the link will rot and not provide any information as to how old the port is. Please include a link to the upstream source code at a particular tag or commit.

There are some updates to `OrderedDictionary<TKey, TValue>` on the main branch that include an overload of `TryAdd()` that outputs the index that I think we should grab. In this case, just use a URL with the git hash code of a recent commit with a comment that these changes will be released in 9.0.6.

### Harmony/JDK Tests

There are also Harmony tests that we should port as `TestOrderedDictionary`.

https://github.com/apache/harmony/blob/trunk/classlib/modules/luni/src/test/api/common/org/apache/harmony/luni/tests/java/util/LinkedHashMapTest.java

We should also get the tests from the JDK. It is okay to nest them inside `TestOrderedDictionary` like was done [here](https://github.com/NightOwl888/J2N/blob/8845a696e2cd3e2ca0752823c779c72aa7f0e5a1/tests/NUnit/J2N.Tests/Collections/Generic/TestLinkedHashSet.cs#L364).

https://github.com/bpupadhyaya/openjdk-8/tree/master/jdk/test/java/util/LinkedHashMap

Since we don't ship the tests to users, there shouldn't be any licensing issue with using them to verify the port works in a way that is similar to Java. But please do include license headers so interested parties can understand what license a particular piece of code is covered under.

File: src/J2N/StaticThrowHelper.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if !FEATURE_STATIC_EXCEPTION_HELPERS
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/J2N/StaticThrowHelper.cs:4]: Naming this as `J2N.Collections.ThrowHelper` has created almost 4,000 compiler errors due to ambiguity with `J2N.ThrowHelper`. I do not think it is worth chasing that down just for purity with upstream. I'm going to leave this named `J2N.Collections.StaticThrowHelper` for now.

File: src/J2N/Collections/Generic/OrderedDictionary.cs [hunk 7]
```diff
@@ -0,0 +1,2108 @@
...
+                }
+
+                return null;
+            }
+            set
+            {
+                // J2N: allow null keys
+                //ThrowIfNull(key, ExceptionArgument.key);
+
+                IfNullAndNullsAreIllegalThenThrow<TValue>(value, ExceptionArgument.value);
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/J2N/Collections/Generic/OrderedDictionary.cs:368]: I thought in an earlier comment that you wanted me to use the IsCompatibleKey approach? I'll add the missing IfNullAndNullsAreIllegalThenThrow line though.

File: src/J2N/Collections/Generic/OrderedDictionary.cs [hunk 8]
```diff
@@ -0,0 +1,2108 @@
...
+using J2N.Runtime.CompilerServices;
+using J2N.Text;
+
+#if FEATURE_STATIC_EXCEPTION_HELPERS
+using static System.ArgumentNullException;
+using static System.ArgumentOutOfRangeException;
+#else
+using static J2N.StaticThrowHelper;
+#endif
+using static J2N.ThrowHelper;
```
Comment (NightOwl888) [inline src/J2N/Collections/Generic/OrderedDictionary.cs]: Please do not use a static import for `J2N.ThrowHelper`. The only reason why we are using implicit methods is to conditionally compile. The rest should be explicit so we don't create ambiguity where the methods are located.

**Review by NightOwl888 — CHANGES_REQUESTED** [review]
Please configure the J2N.Collections.Generic.OrderedDictionary.Tests project in `publish-test-results-for-test-projects.yml` for the results to be published to Azure DevOps.

https://github.com/paulirwin/J2N/blob/issue/11/.build/azure-templates/publish-test-results-for-test-projects.yml#L135

File: tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+
+        #endregion
+
+        #region TryAdd
+        [Fact]
+        public void TryAdd_NullKeyDoesNotThrow() // J2N: allow null keys, was TryAdd_NullKeyThrows
+        {
+            if (default(TKey) is not null)
+            {
+                return;
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.cs:154]: Respectfully, this makes no sense. We are purposefully creating this type to _not_ throw if the key is null, and allow null keys. If I add that back in as you say, then _the tests fail_ because _we do not throw if the key is null_. I am beyond confused by this comment.

I will add the overload tests, though.

File: tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.Values.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        protected override bool NullAllowed => true;
+        protected override bool DuplicateValuesAllowed => true;
+        protected override bool IsReadOnly => true;
+        protected override bool Enumerator_Empty_UsesSingletonInstance => true;
+        protected override bool Enumerator_Empty_Current_UndefinedOperation_Throw => true;
+        protected override IEnumerable<ModifyEnumerable> GetModifyEnumerables(ModifyOperation operations) => new List<ModifyEnumerable>();
+        protected override bool SupportsSerialization => false;
+        protected override Type ICollection_NonGeneric_CopyTo_ArrayOfEnumType_ThrowType => typeof(ArgumentException);
+
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply tests/Xunit/J2N.Collections.Generic.OrderedDictionary.Tests/OrderedDictionary.Generic.Tests.Values.cs:73]: Disregard, I just saw your other comment.

**Review by NightOwl888 — APPROVED** [review]
After a small amount of cleanup and porting the tests from Apache Harmony and JDK8, this looks good.

The `net47` tests (for the `net40` target) failed locally, but succeeded in CI for some reason. I checked and we are definitely testing the correct J2N binary for `net40` and we are getting reports in Azure DevOps for the tests, but somehow it was allowing `IReadOnlyDictionary<TKey, TValue>` to be cast which didn't cause the failure. This doesn't seem like an issue worth pursuing any further, though.

Provided you don't find any issues with the changes I made, this is good to merge.

I added #139 to track adding the functionality to allow deleting while iterating forward. We will also need to add issues in Lucene.NET, ICU4N, and Morfologik.Stemming to replace `LinkedDictionary<TKey, TValue>` with `OrderedDictionary<TKey, TValue>` once we have a release with this included.
